// @wareraprojects/api/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  createAPIClient: () => createAPIClient,
  createTrpcLikeClient: () => createAPIClient
});
module.exports = __toCommonJS(index_exports);

// src/trpc-client.ts
var import_client = require("@trpc/client");
function parseCursorDate(cursor) {
  if (!cursor || typeof cursor !== "string") return null;
  const pipeIndex = cursor.indexOf("|");
  if (pipeIndex === -1) return null;
  const dateStr = cursor.substring(0, pipeIndex);
  if (dateStr === "undefined") return null;
  try {
    const date = new Date(dateStr);
    return isNaN(date.getTime()) ? null : date;
  } catch {
    return null;
  }
}
async function* autoPaginate(client, path, input, options) {
  let currentCursor = input.cursor;
  let pageCount = 0;
  const maxPages = options.maxPages ?? Infinity;
  while (pageCount < maxPages) {
    const requestInput = currentCursor ? { ...input, cursor: currentCursor } : input;
    const response = await client.query(path, requestInput);
    yield {
      items: response.items,
      cursor: response.nextCursor || ""
    };
    pageCount++;
    if (!response.nextCursor || response.items.length === 0 || response.nextCursor.includes("undefined")) {
      break;
    }
    if (options.cursorEnd) {
      const cursorDate = parseCursorDate(response.nextCursor);
      if (cursorDate && cursorDate < options.cursorEnd) {
        break;
      }
    }
    currentCursor = response.nextCursor;
  }
}
function createRateLimitedFetch(origFetch, rateLimit = 100) {
  const f = origFetch ?? globalThis.fetch;
  const delayMs = Math.max(1, Math.floor(6e4 / rateLimit));
  const queue = [];
  let running = false;
  let lastTime = 0;
  const runNext = async () => {
    if (queue.length === 0) {
      running = false;
      return;
    }
    running = true;
    const now = Date.now();
    const elapsed = now - lastTime;
    const wait = Math.max(0, delayMs - elapsed);
    if (wait > 0) await new Promise((res) => setTimeout(res, wait));
    const item = queue.shift();
    lastTime = Date.now();
    f(item.args[0], item.args[1]).then(item.resolve).catch(item.reject).finally(() => {
      if (queue.length > 0) {
        setTimeout(runNext, 0);
      } else {
        running = false;
      }
    });
  };
  return ((input, init) => new Promise((resolve, reject) => {
    queue.push({ args: [input, init], resolve, reject });
    if (!running) runNext();
  }));
}
function createPostQueryFetch(origFetch) {
  const f = origFetch ?? globalThis.fetch;
  return ((input, init) => {
    const isRequest = typeof Request !== "undefined" && input instanceof Request;
    const method = (init?.method ?? (isRequest ? input.method : "GET")).toUpperCase();
    if (method !== "GET") {
      return f(input, init);
    }
    const urlString = typeof input === "string" || input instanceof URL ? input.toString() : input.url;
    let urlObj;
    try {
      urlObj = new URL(urlString);
    } catch {
      return f(input, init);
    }
    const inputParam = urlObj.searchParams.get("input");
    if (!inputParam) {
      return f(input, init);
    }
    urlObj.searchParams.delete("input");
    const headers = new Headers(isRequest ? input.headers : void 0);
    if (init?.headers) {
      new Headers(init.headers).forEach((value, key) => {
        headers.set(key, value);
      });
    }
    if (!headers.has("content-type")) {
      headers.set("content-type", "application/json");
    }
    const nextInit = {
      ...init,
      method: "POST",
      headers,
      body: inputParam
    };
    return f(urlObj.toString(), nextInit);
  });
}
function createBatchLoggingFetch(origFetch, logBatches) {
  if (!logBatches) return origFetch;
  const logFn = typeof logBatches === "function" ? logBatches : (info) => {
    console.log("[trpc-batch]", info);
  };
  return ((input, init) => {
    const urlString = typeof input === "string" || input instanceof URL ? input.toString() : input.url;
    try {
      const urlObj = new URL(urlString);
      const path = urlObj.pathname;
      const lastSegment = path.split("/").pop() ?? "";
      const paths = lastSegment.split(",").filter(Boolean);
      const isBatch = urlObj.searchParams.get("batch") === "1" || paths.length > 1;
      if (isBatch) {
        const body = typeof init?.body === "string" ? init.body : void 0;
        logFn({
          method: (init?.method ?? input.method ?? "GET").toUpperCase(),
          url: urlObj.toString(),
          path: lastSegment,
          paths,
          batchSize: paths.length,
          body
        });
      }
    } catch {
    }
    return origFetch(input, init);
  });
}
function createAPIClient(options) {
  const appliedRateLimit = options?.rateLimit ?? (options?.apiKey !== void 0 ? 200 : 100);
  const baseFetch = options?.fetch ?? globalThis.fetch;
  const loggedFetch = createBatchLoggingFetch(baseFetch, options?.logBatches);
  const postQueryFetch = createPostQueryFetch(loggedFetch);
  const rateLimitedFetch = createRateLimitedFetch(postQueryFetch, appliedRateLimit);
  const client = (0, import_client.createTRPCUntypedClient)({
    links: [
      ...options?.logger === true ? [(0, import_client.loggerLink)()] : [],
      (0, import_client.httpBatchLink)({
        url: options?.url ?? "https://api2.warera.io/trpc",
        fetch: rateLimitedFetch,
        maxURLLength: 16e3,
        headers() {
          return {
            ...options?.headers ?? {},
            ...options?.apiKey ? { "x-api-key": options.apiKey } : {}
          };
        }
      })
    ]
  });
  const makeProxy = (parts) => new Proxy(() => {
  }, {
    get(_t, prop) {
      if (typeof prop !== "string") return void 0;
      if (prop === "_ce" && parts.length === 0) {
        return () => makeProxy([]);
      }
      return makeProxy([...parts, prop]);
    },
    apply(_t, _thisArg, argArray) {
      const path = parts.join(".");
      const input = argArray?.[0] ?? {};
      if (input.autoPaginate === true) {
        const { autoPaginate: _unused, maxPages, cursorEnd, ...cleanedInput } = input;
        return autoPaginate(client, path, cleanedInput, {
          maxPages,
          cursorEnd
        });
      }
      return client.query(path, input);
    }
  });
  return makeProxy([]);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createAPIClient,
  createTrpcLikeClient
});
